packages/sync/src/firehose/index.ts

```diff
@@ -0,0 +1,342 @@
+            for (const write of parsed) {
+              try {
+                await this.opts.handleEvt(write)
+              } catch (err) {
+                this.opts.onError(new FirehoseHandlerError(err, write))
+              }
+            }
+          })
+        } else {
+          await this.processEvt(evt)
+        }
+      }
+    } catch (err) {
+      if (err && err['name'] === 'AbortError') {
+        this.destoryDefer.resolve()
+        return
+      }
+      this.opts.onError?.(new FirehoseSubscriptionError(err))
+      await wait(3000)
+      return this
+    }
+  }
+
+  private async parseEvt(evt: RepoEvent): Promise<Event[]> {
+    try {
+      if (isCommit(evt) && !this.opts.excludeCommit) {
+        const parsed = this.opts.unauthenticatedCommits
+          ? await parseCommitUnauthenticated(evt)
+          : await parseCommitAuthenticated(this.opts.idResolver, evt)
+        const filtered = parsed.filter(
+          (write) =>
+            !this.opts.filterCollections ||
+            this.opts.filterCollections.includes(write.uri.collection),
+        )
+        return filtered
+      } else if (isAccount(evt) && !this.opts.excludeAccount) {
+        const parsed = parseAccount(evt)
+        return parsed ? [parsed] : []
+      } else if (isIdentity(evt) && !this.opts.excludeIdentity) {
+        const parsed = await parseIdentity(
+          this.opts.idResolver,
+          evt,
+          this.opts.unauthenticatedHandles,
+        )
+        return parsed ? [parsed] : []
+      } else {
+        return []
+      }
+    } catch (err) {
+      this.opts.onError?.(new FirehoseParseError(err, evt))
+      return []
+    }
+  }
+
+  private async processEvt(evt: RepoEvent) {
+    const parsed = await this.parseEvt(evt)
+    for (const write of parsed) {
+      try {
+        await this.opts.handleEvt(write)
+      } catch (err) {
+        this.opts.onError(new FirehoseHandlerError(err, write))
+      }
+    }
+  }
+
+  async destroy(): Promise<void> {
+    this.abortController.abort()
+    await this.destoryDefer.complete
+  }
+}
+
+export const parseCommitAuthenticated = async (
+  idResolver: IdResolver,
+  evt: Commit,
+  forceKeyRefresh = false,
+): Promise<CommitEvt[]> => {
+  const did = evt.repo
+  const key = await idResolver.did.resolveAtprotoKey(did, forceKeyRefresh)
+  const claims = evt.ops.map((op) => {
+    const { collection, rkey } = parseDataKey(op.path)
+    return {
+      collection,
+      rkey,
+      cid: op.action === 'delete' ? null : op.cid,
+    }
+  })
+  const verifiedCids: Record<string, CID | null> = {}
+  try {
+    const results = await verifyProofs(evt.blocks, claims, did, key)
+    results.verified.forEach((op) => {
+      const path = formatDataKey(op.collection, op.rkey)
+      verifiedCids[path] = op.cid
+    })
+  } catch (err) {
+    if (err instanceof RepoVerificationError) {
+      return parseCommitAuthenticated(idResolver, evt, true)
+    }
+    throw err
```

Comment:
Probably want to ensure this retries just once, wont loop.
```suggestion
    if (err instanceof RepoVerificationError && !forceKeyRefresh) {
      return parseCommitAuthenticated(idResolver, evt, true)
    }
    throw err
```

Reply:
yeah good call